### PR TITLE
fixes

### PR DESCRIPTION
**Pull request by svlachakis**
#### What types of changes does your code introduce?

- [ ] Bugfix (a non-breaking change that fixes an issue)
- [ ] New feature (a non-breaking change that adds functionality)
- [ ] Breaking change (a change that causes existing functionality not to work as expected)
- [ ] Optimization
- [x] Refactoring
- [ ] Documentation update
- [ ] Build-related changes
- [ ] Other: _Description_

#### Requires testing

- [x] Yes
- [ ] No

#### If yes, did you write tests?

- [x] Yes
- [ ] No

### CODE REVIEW

**Comment by svlachakis** [comment]
@claude review

**Comment by svlachakis** [comment]
@claude review again

**Comment by svlachakis** [comment]
@claude changed to timeout approach, I think it's better. Also for your other comment, When the timeout fires, Engine.Interrupt() causes step() to throw without invoking fault(). Geth exhibits the same behaviour on timeout (evm.Cancel skips fault), so this should be fine shouldn't it?

**Comment by svlachakis** [comment]
@claude review again, comment for V8MaxOldSpaceMb is not needed

**Comment by svlachakis** [comment]
@claude review again

**Comment by svlachakis** [comment]
@claude review again, this test is not worth it and it's flaky

**Comment by svlachakis** [comment]
@claude review again

File: src/Nethermind/Nethermind.Blockchain/Tracing/GethStyle/GethTraceOptions.cs [hunk 1]
```diff
@@ -21,7 +21,11 @@ public record GethTraceOptions
 
     public bool DisableStack { get; init; }
 
-    public string Timeout { get; init; }
+    /// <summary>
+    /// Maximum duration for a single trace. Uses .NET TimeSpan format, e.g. "00:00:05" for 5 seconds.
+    /// Defaults to 5 seconds when null or empty. Throws <see cref="ArgumentException"/> for invalid values.
+    /// </summary>
+    public string? Timeout { get; init; }
```
Comment (LukaszRozmej) [inline src/Nethermind/Nethermind.Blockchain/Tracing/GethStyle/GethTraceOptions.cs]: Is this format compatible with Geth?

File: src/Nethermind/Nethermind.Blockchain/Tracing/GethStyle/Custom/JavaScript/GethLikeJavaScriptTxTracer.cs [hunk 2]
```diff
@@ -55,6 +60,14 @@ public GethLikeJavaScriptTxTracer(
         {
             _tracer.setup(options.TracerConfig?.ToString() ?? "{}");
         }
+
+        TimeSpan timeout = string.IsNullOrEmpty(options.Timeout)
+            ? DefaultTimeout
+            : TimeSpan.TryParse(options.Timeout, out TimeSpan parsed)
+                ? parsed
```
Comment (LukaszRozmej) [inline src/Nethermind/Nethermind.Blockchain/Tracing/GethStyle/Custom/JavaScript/GethLikeJavaScriptTxTracer.cs]: should parsed format have some kind of cap?

**Comment by svlachakis** [comment]
@claude resolved @LukaszRozmej's comments, added geth format support alongside c# time format support. review again please

**Comment by LukaszRozmej** [comment]
## CustomTimeDurationConverter — refactor notes (commit e72b6fd4e4)

### Zero-allocation JSON paths
- **`Read`**: `Utf8JsonReader.CopyString` into a 64-char stack buffer instead of `reader.GetString()` (heap alloc). Uses `HasValueSequence ? ValueSequence.Length : ValueSpan.Length` — UTF-8 byte count is a safe upper bound on UTF-16 char count. The `stackalloc` lives in its own helper (`ReadFromStackBuffer`) so the buffer is only reserved and zero-initialized on the path that actually uses it; fast exits (null token, byteLen > 64) pay nothing. Fallback to `GetString()` for the rare oversized case.
- **`Write`**: `Format` now writes into a caller-provided `Span<char>` and returns `ReadOnlySpan<char>` fed directly into `Utf8JsonWriter.WriteStringValue(ReadOnlySpan<char>)` — no `new string(buf[w..])` allocation. Same helper pattern as `Read` (`WriteFormatted`) so the format buffer is only reserved when `value is not null`.

### Unit dictionary → `SearchValues` + switch expression
- Replaced `Dictionary<string, ulong> UnitMap` (hashing + string allocation for the unit substring) with `SearchValues<char> UnitTerminators = SearchValues.Create(".0123456789")` + a switch expression over `ReadOnlySpan<char>`. No allocations, no hashing, no `s[unitStart..i]` substring.
- Unit-end scan `while (i < s.Length && s[i] != '.' && !IsDigit(s[i])) i++;` → `s.IndexOfAny(UnitTerminators)` (SIMD-capable one-liner).

### Correctness fix in integer overflow check
Previous guard was `v > ulong.MaxValue / 10` before the multiply, which allowed `v * 10 + digit` to wrap when the digit was 6-9 (since `(ulong.MaxValue / 10) * 10 + 9` overflows past `ulong.MaxValue`). Now uses Go's `1<<63` cap with a post-check — matches Go exactly.

### Fractional math: `double` → `UInt128`
Go uses `float64 * (float64 / scale)` with a comment noting it must be nanosecond-accurate for fractions of an hour. Switched to `(UInt128)f * unitNs / scale` — exact integer arithmetic, no rounding, same result for representable values.

### Span-first parsing
- `TryParseGoDuration` takes `ReadOnlySpan<char>` instead of `string`; segments are consumed by slicing (`s = s[1..]`) rather than `substring`.
- Digit runs are bounded with `s.IndexOfAnyExcept(Digits)` (SIMD-capable) via a `TakeDigits` helper, so accumulation loops get clean slices to iterate over.
- Integer part: `ulong.TryParse` with built-in overflow detection, then one `v <= MaxNs` bound check.
- Fractional part: cap at 18 digits (always fits `ulong`) and look up `scale` via a `Pow10` table. Go accumulates 18-19 digits depending on value; we always truncate at 18. The difference is below the 100 ns tick boundary.

### Readability cleanups
- Per-segment parsing split into `TryParseSegment` + `TryTakeInteger` / `TakeFraction` / `TryTakeUnit` / `TakeDigits` helpers — each does one thing.
- Named constants replace magic numbers: `NsPerTick = 100`, `SecondsPerMinute = 60`, `MinutesPerHour = 60`, `NsDigitsInMicrosecond/Millisecond/Second = 3/6/9`, `FormatBufferSize = 32`, `ReadStackBufferSize = 64`, `MaxFractionalDigits = 18`, `MaxNs = 1UL << 63`.
- Sub-second unit/precision selection in `Format` collapsed to a tuple `switch` on `u`.
- Consistent `buf[--w] = c` (replaces split `w--; buf[w] = c`).
- `result = default` → `result = TimeSpan.Zero`.
- `TakeFraction` captures `ulong.TryParse` return with a `Debug.Assert` to document that parsing cannot fail on ASCII-digit input capped at 18 chars.

### 18 new test cases
Added edge cases targeting paths the refactor touched:

**Valid parse (6)**: `"1.s"` (trailing dot, no fraction), `".5s"` (leading dot, no integer), `"0.5h"` (fractional × hour — the `UInt128` path), `"0.123456789s"` (9-digit precision), `"-1h30m"` (negative multi-segment), `"05s"` (leading zeros — the `ulong.TryParse` change).

**Invalid parse (4)**: `"+"`, `"1.2.3s"`, `"5 s"`, `"1ss"`.

**Write (5)**: `FromMilliseconds(-1500)` → `"-1.5s"`, `FromMicroseconds(-500)` → `"-500µs"`, `FromTicks(1)` → `"100ns"`, `FromHours(1)` → `"1h0m0s"`, `FromSeconds(60)` → `"1m0s"`.

**Round-trip (3)**: negative, `FromMicroseconds(1500)` (µs → ms cross-unit), `FromHours(24)`.

**Verification**: ran all 60 tests against both the pre-refactor and post-refactor implementations — both pass all of them, confirming the refactor is behavior-preserving across every edge case we identified.

**Comment by LukaszRozmej** [comment]
@claude re-review

**Comment by svlachakis** [comment]
@LukaszRozmej Thanks, LGTM!

**Review by LukaszRozmej — APPROVED** [review]

**Comment by LukaszRozmej** [comment]
@svlachakis please merge master

**Review by smartprogrammer93 — APPROVED** [review]

**Comment by svlachakis** [comment]
@claude I did some more refactoring please check.

**Comment by svlachakis** [comment]
@LukaszRozmej I did some more refactoring on the go converter, please check again

**Review by flcl42 — APPROVED** [review]
